src/models/Integration.php

```diff
@@ -0,0 +1,379 @@
+        $team_id = await self::genRegisterTeam($email, $id);
+        if (is_int($team_id) === true) {
+          $set_integrations = await self::genSetTeamIntegrations(
+            $team_id,
+            'google_oauth',
+            $email,
+            $id,
+          );
+          if ($set_integrations === true) {
+            $url = await self::genLoginURL('google_oauth', $email);
+          }
+        }
+      }
+    } else if ($error !== false) {
+      $url = '/index.php?page=login';
+    }
+
+    return $url;
+  }
+
+  public static async function genLoginURL(
+    string $type,
+    string $token,
+  ): Awaitable<string> {
+    $team = await Team::genTeamFromOAuthToken($type, $token);
+
+    SessionUtils::sessionRefresh();
+    if (!SessionUtils::sessionActive()) {
+      SessionUtils::sessionSet('team_id', strval($team->getId()));
+      SessionUtils::sessionSet('name', $team->getName());
+      SessionUtils::sessionSet(
+        'csrf_token',
+        (string) gmp_strval(
+          gmp_init(bin2hex(openssl_random_pseudo_bytes(16)), 16),
+          62,
+        ),
+      );
+      SessionUtils::sessionSet(
+        'IP',
+        must_have_string(Utils::getSERVER(), 'REMOTE_ADDR'),
+      );
+      if ($team->getAdmin()) {
+        SessionUtils::sessionSet('admin', strval($team->getAdmin()));
+      }
+    }
+    if ($team->getAdmin()) {
+      $redirect = 'admin';
+    } else {
+      $redirect = 'game';
+    }
+
+    $login_url = '/index.php?p='.$redirect;
+
+    return $login_url;
+  }
+
+  public static async function genFacebookOAuth(): Awaitable<bool> {
+    list($client, $url) = await self::genFacebookAuthURL("oauth");
+    $helper = $client->getRedirectLoginHelper();
+
+    $code = idx(Utils::getGET(), 'code');
+    $error = idx(Utils::getGET(), 'error');
+
+    if (!is_string($code)) {
+      $code = false;
+    }
+
+    if (!is_string($error)) {
+      $error = false;
+    }
+
+    $accessToken = '';
+
+    if ($code !== false) {
+      $graph_error = false;
+      try {
+        $accessToken = $helper->getAccessToken();
+      } catch (Facebook\Exceptions\FacebookResponseException $e) {
+        $graph_error = true;
+      } catch (Facebook\Exceptions\FacebookSDKException $e) {
+        $graph_error = true;
+      }
+
+      if ($graph_error === true) {
+        return false;
+      } else {
+        $response =
+          $client->get('/me?fields=id,third_party_id,email', $accessToken);
+        $profile = $response->getGraphUser();
+        $email = $profile['email'];
+        $id = $profile['third_party_id'];
+
+        $set_integrations = await self::genSetTeamIntegrations(
+          SessionUtils::sessionTeam(),
+          'facebook_oauth',
+          $email,
+          $id,
+        );
+        return $set_integrations;
+      }
+    } else if ($error !== false) {
+      return false;
+    }
+
+    header('Location: '.filter_var($url, FILTER_SANITIZE_URL));
+    exit;
+    return false;
```

Comment:
Is this required? The typechecker should realize it's unreachable

Reply:
Seems to be.  We were getting an error from the typechecker, and we didn't want to make the method nullable.